src/inc/til/throttled_func.h

```diff
@@ -0,0 +1,206 @@
+        // Since the timer cannot be recreated, instances cannot be moved either.
+        throttled_func(const throttled_func&) = delete;
+        throttled_func& operator=(const throttled_func&) = delete;
+        throttled_func(throttled_func&&) = delete;
+        throttled_func& operator=(throttled_func&&) = delete;
+
+        // Throttles the invocation of the function passed to the constructor.
+        // If this is a trailing_throttled_func:
+        //   If you call this function again before the underlying
+        //   timer has expired, the new arguments will be used.
+        template<typename... MakeArgs>
+        void operator()(MakeArgs&&... args)
+        {
+            if (!_storage.emplace(std::forward<MakeArgs>(args)...))
+            {
+                _leading_edge();
+            }
+        }
+
+        // Modifies the pending arguments for the next function
+        // invocation, if there is one pending currently.
+        //
+        // `func` will be invoked as func(Args...). Make sure to bind any
+        // arguments in `func` by reference if you'd like to modify them.
+        template<typename F>
+        void modify_pending(F func)
+        {
+            _storage.modify_pending(func);
+        }
+
+        // Makes sure that the currently pending timer is executed
+        // as soon as possible and in that case waits for its completion.
+        // You can use this function in your destructor to ensure that any
+        // pending callback invocation is completed as soon as possible.
+        //
+        // NOTE: Don't call this function if the operator()
+        //       could still be called concurrently.
+        void flush()
+        {
+            WaitForThreadpoolTimerCallbacks(_timer.get(), true);
+            if (_storage)
+            {
+                _trailing_edge();
+            }
+        }
+
+    private:
+        static void __stdcall _timer_callback(PTP_CALLBACK_INSTANCE /*instance*/, PVOID context, PTP_TIMER /*timer*/) noexcept
+        try
+        {
+            static_cast<throttled_func*>(context)->_trailing_edge();
+        }
+        CATCH_LOG()
+
+        void _leading_edge()
+        {
+            if constexpr (leading)
+            {
+                _func();
+            }
+
+            SetThreadpoolTimerEx(_timer.get(), &_delay, 0, 0);
+        }
+
+        void _trailing_edge()
+        {
+            if constexpr (leading)
+            {
+                _storage.reset();
+            }
+            else
+            {
+                std::apply(_func, _storage.take());
+            }
+        }
+
+        inline wil::unique_threadpool_timer _createTimer()
+        {
+            wil::unique_threadpool_timer timer{ CreateThreadpoolTimer(&_timer_callback, this, nullptr) };
+            THROW_LAST_ERROR_IF(!timer);
+            return timer;
+        }
+
+        FILETIME _delay;
+        function _func;
+        wil::unique_threadpool_timer _timer;
```

Comment:
Does the destruction of a unique_threadpool_timer deschedule it?

Reply:
Yep! The destructor of `wil::unique_threadpool_timer` calls `WaitForThreadpoolTimerCallbacks(_timer, true)` and thus cancels any pending timers.